### PR TITLE
V2

### PR DESCRIPTION
**Pull request by nextdayy**
Poly UI V2

hopes:
- [X] Fully automatic UI positioning system which works on nearby elements, for example you can create a button just by saying ‘I want a block with an image and text in it’ and nothing more
- [X] Additionally on a higher-level positioning of components, etc. will work in a similar way
- [X] Allow for potentially complex ways of positioning, such as percentage based sizing and positioning
- [ ] Allow for multiple ways to create UIs without necessarily using Kotlin syntax
- [X] ~~Still have the flexibility that properties were supposed to bring~~, and the power of the events system

The general idea and premise:
- [X] the current positioning system is very stupid and hardly works in complex situations
- [X] So, remove it entirely and separate out the positioning of the UI from the rest of the systems
- [X] Remove layouts and allow components to potentially contain other components, 
- [ ] where frame buffers will still be used (note the benefits of frame buffers from testing have been much smaller than initially believed)
- [X] Remove properties as they were a bit of a nightmare to use
- [X] Essentially make it so that the way that the UI is positioned is handled entirely separately from the operation of the UI, and this process is run once at startup
- [X] Make drawable operations more useful, and make it so more things must follow stricter rules (such as event handling where there was 2 ways of doing it), moving recolouring into a drawable operation, and keyframes
- [X] As a consequence, separate out the frontend syntax from created components allowing for many different ways of creating a UI to be done and new ways of positioning to be added
- [X] Refactor events to be more structured
- [X] Input system is good, along with render backend

Potential issues
- [ ] Runtime movement of elements - the system will be equipped with features which will reposition elements if needed, such as resetting to how they were, or moving overlapping elements away from each other
- [X] Scrolling: groups of components will be able to be scrolled

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
-version=0.25.0
+version=2.0.102
```
Comment (Wyvest) [inline gradle.properties:1]: What is this new versioning bro

File: nanovg-impl/src/main/kotlin/org/polyfrost/polyui/renderer/impl/GLWindow.kt [hunk 2]
```diff
@@ -222,7 +221,14 @@ class GLWindow @JvmOverloads constructor(
             polyUI.eventManager.keyTyped(codepoint.toChar())
         }
 
+        var ran = false
         glfwSetScrollCallback(handle) { _, x, y ->
+            // asm: small scroll amounts are usually trackpads
+            if (!ran && (y < 1.0 && x < 1.0) && PolyUI.isOnMac && !polyUI.settings.naturalScrolling) {
```
Comment (Wyvest) [inline nanovg-impl/src/main/kotlin/org/polyfrost/polyui/renderer/impl/GLWindow.kt:227]: Does this actually work

**Review by Wyvest — APPROVED** [review]
lgtm

File: gradle.properties (hunk 1 above)
Comment (nextdayy) [reply gradle.properties:1]: well its v2 plus some as yes

File: nanovg-impl/src/main/kotlin/org/polyfrost/polyui/renderer/impl/GLWindow.kt (hunk 2 above)
Comment (nextdayy) [reply nanovg-impl/src/main/kotlin/org/polyfrost/polyui/renderer/impl/GLWindow.kt:227]: yes (i only code on a trackpad now)

**Review by pauliesnug — APPROVED** [review]
lgtm- i will look into kotlinter rn

File: gradle.properties (hunk 1 above)
Comment (pauliesnug) [reply gradle.properties:1]: maybe make a new versioning semantics page for polyui in contributor docs

**Pull request merged**